kedro/config/omegaconf_config.py

```diff
@@ -0,0 +1,258 @@
+
+    """
+
+    def __init__(
+        self,
+        conf_source: str,
+        env: str = None,
+        runtime_params: Dict[str, Any] = None,
+        config_patterns: Dict[str, List[str]] = None,
+        *,
+        base_env: str = "base",
+        default_run_env: str = "local",
+    ):
+        """Instantiates a ``OmegaConfLoader``.
+
+        Args:
+            conf_source: Path to use as root directory for loading configuration.
+            env: Environment that will take precedence over base.
+            runtime_params: Extra parameters passed to a Kedro run.
+            config_patterns: Regex patterns that specify the naming convention for configuration
+                files so they can be loaded. Can be customised by supplying config_patterns as
+                in `CONFIG_LOADER_ARGS` in `settings.py`.
+            base_env: Name of the base environment. Defaults to `"base"`.
+                This is used in the `conf_paths` property method to construct
+                the configuration paths.
+            default_run_env: Name of the default run environment. Defaults to `"local"`.
+                Can be overridden by supplying the `env` argument.
+        """
+        self.base_env = base_env
+        self.default_run_env = default_run_env
+
+        self.config_patterns = {
+            "catalog": ["catalog*", "catalog*/**", "**/catalog*"],
+            "parameters": ["parameters*", "parameters*/**", "**/parameters*"],
+            "credentials": ["credentials*", "credentials*/**", "**/credentials*"],
+            "logging": ["logging*", "logging*/**", "**/logging*"],
+        }
+        self.config_patterns.update(config_patterns or {})
+
+        super().__init__(
+            conf_source=conf_source,
+            env=env,
+            runtime_params=runtime_params,
+        )
+
+    def __getitem__(self, key) -> Dict[str, Any]:
+        """Get configuration files by key, load and merge them, and
+        return them in the form of a config dictionary.
+
+        Args:
+            key: Key of the configuration type to fetch.
+
+        Raises:
+            KeyError: If key provided isn't present in the config_patterns of this
+               OmegaConfLoader instance.
+            MissingConfigException: If no configuration files exist matching the patterns
+                mapped to the provided key.
+
+        Returns:
+            Dict[str, Any]:  A Python dictionary with the combined
+               configuration from all configuration files.
+        """
+
+        # Allow bypassing of loading config from patterns if a key and value have been set
+        # explicitly on the ``OmegaConfLoader`` instance.
+        if key in self:
+            return super().__getitem__(key)
+
+        # In the first iteration of the OmegaConfLoader we'll keep the resolver turned-off.
+        # It's easier to introduce them step by step, but removing them would be a breaking change.
+        self._clear_omegaconf_resolvers()
```

Comment:
I wonder whether we could call this on `OmegaConfLoader.__init__` once, instead on each request for config. WDYT the implications of this could be?

Reply:
Yes actually I think that makes more sense. Otherwise you could get a situation where if the `__getitem__` function has been called users could technically use the resolvers. Seems unlikely, but I think it's better to move it up to `__init__`.